### PR TITLE
Test that state channel closes within grace blocks of an epoch end are rewarded

### PR DESCRIPTION
**Pull request by mrallen1**
Companion test for https://github.com/helium/blockchain-core/pull/627

### CODE REVIEW

**Comment by mrallen1** [comment]
From the log files:

```
*** User 2020-10-08 21:16:32.668 ***
height post packets 8

*** User 2020-10-08 21:16:50.319 ***
close height 25

*** User 2020-10-08 21:16:50.319 ***
saw closed

[more omitted]

*** User 2020-10-08 21:17:33.501 ***
DCRewards: [19435]
```

File: test/miner_state_channel_SUITE.erl [hunk 1]
```diff
@@ -27,7 +27,8 @@
          conflict_test/1,
          reject_test/1,
          server_doesnt_close_test/1,
-         client_reports_overspend_test/1
+         client_reports_overspend_test/1,
+         dc_rewards_v4_test/1
```
Comment (vihu) [inline test/miner_state_channel_SUITE.erl:32]: Can we rename this test? It's guarded behind v5 now. I'd suggest `missing_dc_reward_for_epoch`...

File: test/miner_ct_utils.erl [hunk 2]
```diff
@@ -1049,6 +1049,13 @@ get_txn([{_, B}], PredFun) ->
                     end,
                     blockchain_block:transactions(B))).
 
+get_txns(Details, PredFun) when is_list(Details) ->
```
Comment (vihu) [inline test/miner_ct_utils.erl:1052]: `Details` -> `Blocks`

File: test/miner_state_channel_SUITE.erl (hunk 1 above)
Comment (mrallen1) [reply test/miner_state_channel_SUITE.erl:31]: sure

File: test/miner_state_channel_SUITE.erl [hunk 3]
```diff
@@ -1042,3 +1131,35 @@ check_type_sc_open(T) ->
...
+       securities_percent => 0.34
+     }.
+
+get_reward_txns(Node) ->
+    Chain = ct_rpc:call(Node, blockchain_worker, blockchain, []),
+    Blocks = ct_rpc:call(Node, blockchain, blocks, [Chain]),
+
+    lists:flatten(
+        maps:fold(
+            fun(_, B, Acc) ->
```
Comment (vihu) [inline test/miner_state_channel_SUITE.erl:1152]: This function can be tweaked to use `get_txn/2` defined in ct utils...

**Comment by mrallen1** [comment]
tested against master on blockchain-core (with reward_version => 4) and

```
----------------------------------------------------
2020-10-09 19:54:13.171
DCRewards: []


----------------------------------------------------
2020-10-09 19:54:15.383
fin


%%% miner_state_channel_SUITE ==> sc_v2.ensure_dc_reward_during_grace_blocks: FAILED
%%% miner_state_channel_SUITE ==> {test_case_failed,"DC rewards contained no values"}
```

so confirmed the test fails as expected against unpatched blockchain-core.

**Review by vihu — APPROVED** [review]

File: rebar.config [hunk 4]
```diff
@@ -10,7 +10,7 @@
 
 {deps, [
         {blockchain, {git, "https://github.com/helium/blockchain-core.git",
-                      {branch, "master"}}},
+                      {branch, "mra/check-previous-grace-blocks"}}},
```
Comment (vihu) [inline rebar.config:13]: This would need to switch to master when the core branch branch lands

File: test/miner_state_channel_SUITE.erl [hunk 5]
```diff
@@ -1078,6 +1090,86 @@ open_state_channel(Config, ExpireWithin, Amount, OUI) ->
     ct:pal("SC: ~p", [SC]),
     ID.
 
+ensure_dc_reward_during_grace_blocks(Config) ->
+    %% NOTE: This test may give a false positive because the check at the bottom doesn't truly verify that the state channel close is within next epoch start - grace period. However, it should always pass when reward_version=5 implying that the dc_reward would be paid out regardless of whether it was within the grace period or not.
```
Comment (evanmcc) [inline test/miner_state_channel_SUITE.erl:1094]: holy long line batman

**Review by evanmcc — APPROVED** [review]

**Pull request merged**